port/linux/omrvmem.c

```diff
@@ -744,16 +743,87 @@ omrvmem_free_memory(struct OMRPortLibrary *portLibrary, void *address, uintptr_t
 		ret = (int32_t)shmdt(address);
 	}
 
-	if((mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN) && fd != -1) {
+	if((mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN) && (-1 != fd) && (-1 != ret)) {
+		void **leafAddresses = (void **)handle;
+		uintptr_t leafCount = (uintptr_t)leafAddresses[0];
+		Trc_PRT_vmem_omrvmem_free_memory_UpdateLeaves(fd, (void *)leafAddresses, leafCount, address, byteAmount);
 		close(fd);
+		if (NULL != leafAddresses && leafCount != 0) {
+			ret = updateDoubleMappedLeaves(portLibrary, leafAddresses, byteAmount, leafCount);
+		}
 	}
-
 	omrmem_categories_decrement_counters(category, byteAmount);
 
 	Trc_PRT_vmem_omrvmem_free_memory_Exit(ret);
 	return ret;
 }
 
+/**
+ * One consequence of double mapping is that double mapped regions become
+ * shared. Originally the heap is private, therefore, when we are ready to
+ * collect double mapped regions we must "fix" these regions by mapping them
+ * back to private. If we don't update the regions, subsequent allocations
+ * on that region will be considered as shared, and given shared memory
+ * limitations on Linux, that can lead to unpredictable crashes.
+ *
+ * @param OMRPortLibrary	*portLibrary		[in] The portLibrary object
+ * @param uintptr_t		byteAmount		[in] total size required to update
+ * @return true if update is successful and false otherwise
+ */
+int32_t
+updateDoubleMappedLeaves(struct OMRPortLibrary *portLibrary, void **leafAddresses, uintptr_t byteAmount, uintptr_t leafCount)
```

Comment:
I'd avoid using 'leaves' anywhere in PortLib. it's too specific name. but in theory it may be used for something else in future. perhaps DiscontiguousRegions/Areas (in J9/OMR 'region' somewhat implies it's fixed size; 'area' is not very used term)?

when from the context it's clear what we are dealing with it can be shortened to just regions/areas.
for example:
updateDoubleMappedLeaves -> updateDoubleMappedDiscontiguousRegions
leafAddresses, leafCount -> regionAddresses, regionCount